src/cupbearer/models/transformers_hf.py

```diff
@@ -0,0 +1,189 @@
+        return embeddings
+
+
+# TODO: test
+class ClassifierTransformerHF(TransformerBaseHF):
+    def __init__(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizerBase,
+        embed_dim: int,
+        max_length: int,
+        num_classes: int,
+    ):
+        super().__init__(
+            model=model, tokenizer=tokenizer, embed_dim=embed_dim, max_length=max_length
+        )
+        self.num_classes = num_classes
+        self.classifier = nn.Linear(self.embed_dim, self.num_classes)
+
+    def forward(self, x: str | list[str]):
+        # get tokens
+        tokens = self.process_input(x)
+        # get embeddings
+        embeddings = self.get_embeddings(tokens)
+
+        # TODO (odk) se store (doesn't this slow down training?)
+
+        # take mean across non-padded dimensions
+        mask = tokens["input_ids"] != self.tokenizer.pad_token_id
+        mask = mask.unsqueeze(-1)
+        assert mask.shape == tokens["input_ids"] + (1,)
+        assert embeddings.shape == tokens["input_ids"] + (self.embed_dim,)
+        embeddings = embeddings * mask
+        embeddings = embeddings.sum(dim=1) / mask.sum(dim=1)
+
+        # compute logits
+        logits = self.classifier(embeddings)
+        return logits
+
+
+class TamperingPredictionTransformer(TransformerBaseHF):
+    # TODO: factor out token processing, create interface for using tokenizer in dataset
+    def __init__(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizerBase,
+        embed_dim: int,
+        max_length: int,
+        n_sensors: int,
+        sensor_token: str = " omit",
+    ):
+        super().__init__(
+            model=model, tokenizer=tokenizer, embed_dim=embed_dim, max_length=max_length
+        )
+        self.n_sensors = n_sensors
+        self.n_probes = self.n_sensors + 1  # +1 for aggregate measurements
+        self.sensor_token_id = self.get_single_token(sensor_token)
+
+        self.probes = nn.ModuleList(
+            [nn.Linear(self.embed_dim, 1) for _ in range(self.n_probes)]
+        )
+
+    def forward(self, x: str | list[str]):
+        tokens = self.process_input(x)
+        embeddings = self.get_embeddings(tokens)
+        b = embeddings.shape[0]
+
+        # TODO (odk) use store
+
+        # sensor embeddings
+        batch_inds, seq_ids = torch.where(
+            tokens["input_ids"] == self.sensor_token_id
+        )  # TODO: pre-specify that its always 3?
+        sensor_embs = embeddings[batch_inds, seq_ids].reshape(
+            b, self.n_sensors, self.embed_dim
+        )
+        # last token embedding (for aggregate measurement)
+        last_token_ind = tokens["attention_mask"].sum(dim=1) - 1
+        last_embs = embeddings[torch.arange(b), last_token_ind]
+        probe_embs = torch.concat([sensor_embs, last_embs.unsqueeze(dim=1)], axis=1)
+        assert probe_embs.shape == (b, self.n_probes, self.embed_dim)
+        logits = torch.concat(
```

Comment:
Probably simpler to make `self.probes` a single `nn.Linear` with `n_probes` output logits? Though if we ever have non-linear probes, then it actually makes a difference (do we share most of their body or not between measurements), not sure which option we'd want in that case